js/ad_manager_vast.js

```diff
@@ -1195,41 +1232,77 @@ OO.Ads.manager(function(_, $) {
      * @returns {string[]|null} The array of tracking urls associated with the event name. Returns null if no URLs exist.
      */
     var _getTrackingEventUrls = function(adObject, trackingEventName) {
+      adObject = _getVastAdObject(adObject);
       var trackingUrls = null;
+
+      // get tracking urls from both the linear and nonLinear object
+      var linearTrackingUrls = [];
+      var nonLinearTrackingUrls = [];
+
       if (adObject &&
-          adObject.ad &&
-          adObject.ad.data &&
-          adObject.ad.data.tracking &&
-          adObject.ad.data.tracking[trackingEventName] &&
-          adObject.ad.data.tracking[trackingEventName].length > 0) {
-        trackingUrls = adObject.ad.data.tracking[trackingEventName];
+          adObject.linear &&
+          adObject.linear.tracking &&
+          adObject.linear.tracking[trackingEventName] &&
+          adObject.linear.tracking[trackingEventName].length > 0) {
+        linearTrackingUrls = adObject.linear.tracking[trackingEventName];
+      }
+
+      if (adObject &&
+          adObject.nonLinear &&
+          adObject.nonLinear.tracking &&
+          adObject.nonLinear.tracking[trackingEventName] &&
+          adObject.nonLinear.tracking[trackingEventName].length > 0) {
+        nonLinearTrackingUrls = adObject.nonLinear.tracking[trackingEventName];
+      }
+
+      if (!_.isEmpty(linearTrackingUrls) || !_.isEmpty(nonLinearTrackingUrls)) {
+        trackingUrls = linearTrackingUrls.concat(nonLinearTrackingUrls);
       }
+
       return trackingUrls;
     };
 
+    /**
+     * Helper function to retrieve the ad object's ad ID (the value of the <Ad> element's "id" attribute).
+     * Not the "id" our ad manager assigns.
+     * @private
+     * @method Vast#_getAdId
+     * @param {object} adObject The ad metadata
+     * @returns {string|null} The ad's ID attribute. Returns null if the ID does not exist.
+     */
+    var _getAdId = function(adObject) {
+      adObject = _getVastAdObject(adObject);
+      var adId = null;
+      if (adObject &&
+          adObject.id) {
+        adId = adObject.id;
+      }
+      return adId;
+    };
 
     /**
      * Helper function to ping URLs in each set of tracking event arrays.
      * @private
      * @method Vast#_pingTrackingUrls
      * @param {object} urlObject An object with the tracking event names and their
      * associated URL array.
+     * @param {object} adId The ad ID
      */
-    var _pingTrackingUrls = _.bind(function(urlObject) {
+    var _pingTrackingUrls = _.bind(function(urlObject, adId) {
       for (var trackingName in urlObject) {
         if (urlObject.hasOwnProperty(trackingName)) {
           try {
             var urls = urlObject[trackingName];
             if (urls) {
               OO.pixelPings(urls);
-              OO.log("VAST: \"" + trackingName + "\" tracking URLs pinged");
+              OO.log("VAST: \"" + trackingName + "\" tracking URLs pinged for VAST Ad Id: " + adId);
```

Comment:
I think this crashes if adId is undefined. Can you double check and add an if statement, if needed?